validation/validate.py

```diff
@@ -170,17 +170,80 @@ def _score_k_fold(thresholds: np.ndarray,
-                      embeddings2: List[List[float]],
+def _get_target_faces(embeddings1: List[FaceVector],
+                      embeddings2: List[FaceVector],
                       distance_metric: DistanceMetric,
-                      is_match: bool) -> Tuple[List[float], List[float]]:
+                      is_match: bool) -> Tuple[FaceVector, FaceVector]:
     X, Y = zip(*[(emb1, emb2) for emb1 in embeddings1 for emb2 in embeddings2])
     distances = _distance_between_embeddings(X, Y, distance_metric)
     distance_criteria = min if is_match else max
     index, _ = distance_criteria(enumerate(distances), key=lambda x: x[1])
     return X[index], Y[index]
 
 
+def _get_container_metrics(face_vectors: List[List[FaceVector]]) -> Tuple[
+                                                                        int,
+                                                                        int,
+                                                                        float]:
+    num_expected = len(face_vectors)
+    num_missing = sum([1 for i in face_vectors if not i])
+    percentage_missing = 100 * (num_missing / num_expected)
+    return num_expected, num_missing, percentage_missing
+
+
+def _remove_empty_embeddings(args: Namespace,
+                             embeddings: np.ndarray,
+                             labels: np.ndarray) -> Tuple[np.ndarray,
+                                                          np.ndarray]:
+    if args.remove_empty_embeddings:
+        embs_filter = embeddings == np.asarray([[0]*args.embedding_size])
+        empty_indices = np.where(np.all(embs_filter, axis=1))[0]
+        pair_empty_indices = np.asarray([i + 1
+                                         if i % 2 == 0
+                                         else i - 1
+                                         for i in empty_indices])
+        embedding_indices = np.unique(np.concatenate((empty_indices,
+                                                      pair_empty_indices)))
+        embeddings = np.delete(embeddings, embedding_indices, axis=0)
+        label_indices = np.unique(embedding_indices // 2)
+        labels = np.delete(labels, label_indices, axis=0)
+    return embeddings, labels
+
+
+def _prealigned(args: Namespace, num_matches_mismatches: int,
+                face_vectors: List[List[FaceVector]]) -> np.ndarray:
+    if args.prealigned:
+        embeddings = np.asarray([[0] * args.embedding_size
+                                 if not embedding else embedding[0]
+                                 for embedding in face_vectors])
+    else:
+        embeddings = []
+        is_match = False
+        for i, (embs1, embs2) in enumerate(zip(face_vectors[0::2],
+                                               face_vectors[1::2])):
+            if i % num_matches_mismatches == 0:
+                is_match = not is_match
+
+            embs1, embs2 = _get_target_faces(
+                embs1 or [[0] * args.embedding_size],
```

Comment:
If you make this code more obvious and less terse, you can fix the type error:

```py
f1 = face_vectors[0::2]
f2 = face_vectors[1::2]
for i, embs1 in enumerate(f1):
    embs2 = f2[i]
    # ...
```

This will actually reveal another type error, namely that `[[0] * args.embedding_size]` isn't really a `List[FaceVector]` since it's technically `List[List[int]]`. You can fix this using `[[0.] * args.embedding_size]`.